src/ethereum/vm/instructions.py

```diff
@@ -44,21 +68,313 @@ def add(evm: Evm) -> None:
-
+    x, y = pop_multiple(evm.stack, 2)
     val = x.wrapping_add(y)
 
     push(evm.stack, val)
 
 
+def sub(evm: Evm) -> None:
+    """
+    Subtracts the top two elements of the stack, and pushes the result back
+    on the stack.
+
+    Parameters
+    ----------
+    evm : `Evm`
+        The current EVM frame.
+
+    Raises
+    ------
+    StackUnderflowError
+        If `len(stack)` is less than `2`.
+    OutOfGasError
+        If `evm.gas_left` is less than `GAS_VERY_LOW`.
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_VERY_LOW)
+
+    x, y = pop_multiple(evm.stack, 2)
+    val = x.wrapping_sub(y)
+
+    push(evm.stack, val)
+
+
+def mul(evm: Evm) -> None:
+    """
+    Multiply the top two elements of the stack, and pushes the result back
+    on the stack.
+
+    Parameters
+    ----------
+    evm : `Evm`
+        The current EVM frame.
+
+    Raises
+    ------
+    StackUnderflowError
+        If `len(stack)` is less than `2`.
+    OutOfGasError
+        If `evm.gas_left` is less than `GAS_LOW`.
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_LOW)
+
+    x, y = pop_multiple(evm.stack, 2)
+    val = x.wrapping_mul(y)
+
+    push(evm.stack, val)
+
+
+def div(evm: Evm) -> None:
+    """
+    Integer division of the top two elements of the stack. Pushes the result
+    back on the stack.
+
+    Parameters
+    ----------
+    evm : `Evm`
+        The current EVM frame.
+
+    Raises
+    ------
+    StackUnderflowError
+        If `len(stack)` is less than `2`.
+    OutOfGasError
+        If `evm.gas_left` is less than `GAS_LOW`.
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_LOW)
+
+    x, y = pop_multiple(evm.stack, 2)
+    res = (x // y) if y != 0 else U256(0)
+
+    push(evm.stack, res)
+
+
+def sdiv(evm: Evm) -> None:
+    """
+    Signed integer division of the top two elements of the stack. Pushes the
+    result back on the stack.
+
+    Parameters
+    ----------
+    evm : `Evm`
+        The current EVM frame.
+
+    Raises
+    ------
+    StackUnderflowError
+        If `len(stack)` is less than `2`.
+    OutOfGasError
+        If `evm.gas_left` is less than `GAS_LOW`.
+    """
+    evm.gas_left = subtract_gas(evm.gas_left, GAS_LOW)
+
+    x, y = map(
+        lambda u256_element: u256_element.to_signed_int(),
+        pop_multiple(evm.stack, 2),
+    )
+
+    if y == 0:
+        res = 0
+    elif x == -U255_CEIL_VALUE and y == -1:
+        res = -U255_CEIL_VALUE
+    else:
+        sign = signum(x * y)
+        res = sign * (abs(x) // abs(y))
+
+    push(evm.stack, U256.from_signed(res))
+
+
+def mod(evm: Evm) -> None:
+    """
+    Modulo remainder of the top two elements of the stack. Pushes the result
+    back on the stack.
+
+    Parameters
+    ----------
+    evm : `Evm`
+        The current EVM frame.
+
+    Raises
+    ------
+    StackUnderflowError
+        If `len(stack)` is less than `2`.
+    OutOfGasError
+        If `evm.gas_left` is less than `GAS_LOW`.
```

Comment:
I realize it was I who set the pattern of defining the amount of gas that would call a function to fail. I wonder if this is a future foot gun for ourselves trying to keep both up to date... cc @SamWilsn

Reply:
If we run into issues, we can always change it.